### DISCUSSION TITLE
Recursive Proof with v0.9.1

### DISCUSSION THREAD

**Discussion by stephanfeb** [Consensys/gnark#1034] [General]
I noted that the gnark playground has a recursive example using the older API from version 0.8.x. 
The newer 0.9.x APIs use generics, and it was non-obvious to me how I would reconstruct the "Public Inputs" needed to verify the Inner Circuit using the new APIs. 

I have settled on the following bit of code which basically attempts to reconstructs the public portion of the InnerWitness. 

Am I doing this correctly ? 

```
func (circuit *Sha256OuterCircuit[S, G1El, G2El, GtEl]) Define(api frontend.API) error {
.
.  << omitted code >> 
.
	innerCircuit := &Sha256InnerCircuit{
		CurrTxId: circuit.PrevTxId,
	}
	pubWitness, err := frontend.NewWitness(innerCircuit, ecc.BLS12_377.ScalarField())

	witnessVal, err := stdgroth16.ValueOfWitness[S, G1El](pubWitness)

	witness := stdgroth16.Witness[S]{
		Public: witnessVal.Public,
	}

	err = verifier.AssertProof(circuit.VerifyingKey, circuit.InnerProof, witness) // circuit.InnerWitness

       return err
}

```

**Comment by ivokub** [comment]
Have a look at the examples: https://pkg.go.dev/github.com/consensys/gnark@v0.9.2-0.20240205220526-e043c9fe9585/std/recursion/groth16#pkg-overview

Please note that the Groth16 verifier is not complete. Notably due to https://github.com/Consensys/gnark/issues/987 it is not possible to verify all proofs (which use commitments).

**Reply by stephanfeb** [reply]
Does this also hold for Plonk recursion ? I.e. would the recursive verification of a sha256 hash work if I were to switch to the new Plonk verifier ?

**Reply by stephanfeb** [reply]
SOLVED !

This discussion can be closed. I have discovered why my recursive sha256 failed to verify. The key was to ensure that I use `plonk.GetNativeVerifierOptions` and `plonk.GetNativeProverOptions` when computing and verifying the inner proofs. 

code snippet below...

```go
	innerWitness, err := frontend.NewWitness(innerAssignment, field)
	assert.NoError(err)
	innerProof, err := native_plonk.Prove(innerCcs, innerPK, innerWitness, plonk.GetNativeProverOptions(outer, field))
	assert.NoError(err)
	innerPubWitness, err := innerWitness.Public()
	assert.NoError(err)
	err = native_plonk.Verify(innerProof, innerVK, innerPubWitness, plonk.GetNativeVerifierOptions(outer, field))
	assert.NoError(err)
	return innerCcs, innerVK, innerPubWitness, innerProof
```
